File: method.py
import state
import random
import helper
from random import randint
import time
import math
dic = {}

def reproduce(x, y):
    #apply PMX accourding to wikipedia
    size = x.length
    cxpoint1 = randint(0, size)
    l1 =  x.toList()
    l2 = y.toList()
    l3 = list(l1[0:cxpoint1])
    d = dict(dic)
    for i in l3:
        d[i] -= 1
    for i in range(size):
        j = l2[i]
        if d[j] == 0:
            continue
        l3.append(j)
        d[j] -= 1

    return state.State(bin1=l3[0:size//3], bin2=l3[size//3:size//3*2], bin3=l3[size//3*2:])



def genetics(test, size, elite, mutation,ti):
    global dic
    popu = [test]
    dic = test.getdic()

    for i in range(100 - 1):
        popu.append(test.shuffle())
    for i in popu:
        i.setscore()
        
    popu.sort(key=lambda x: x.sc)

    def random_selection():
        length = len(popu)
        choosen = length * (1 - elite)
        c1 = randint(int(choosen), length - 1)
        return popu[c1]


    t = time.time()
    while(time.time() - t < ti):
        new_population = []
        for i in range(size):
            x = random_selection()
            y = random_selection()
            child = reproduce(x, y)
            if helper.decision(mutation):
                child = child.newState()
            new_population.append(child)
            child.setscore()
        popu = new_population
        popu.sort(key=lambda x: x.sc)


    return popu[-1]


def hillclimbing(state, ti):
    potentialSolution = []
    countSolution = 0
    countTime = 0
    currentState = state
    potentialSolution.append(currentState)
    t = time.time()
    while(time.time() - t < ti):
        currentState = potentialSolution.pop()
        nextState = currentState.newState()
        if (nextState.score() > currentState.score()):
            currentState = nextState
            countTime = 0
        else:
            countTime += 1
        potentialSolution.append(currentState)
        if (countTime == 100):
            countTime = 0
            countSolution +=1
            currentState = state.shuffle()
            potentialSolution.append(currentState)

    bestSolution = potentialSolution[0]
    for i in range(0, len(potentialSolution)):
        if (potentialSolution[i].score() > bestSolution.score()):
            bestSolution = potentialSolution[i]

    return bestSolution

def annealing(state, ti, maxtemp, scale):
    potentialSolution = []
    countSolution = 0
    countTime = 0
    countTemp = 0
    currentState = state
    state.setscore()
    potentialSolution.append(currentState)

    t =  time.time()
    while(time.time() - t < ti):
        currentState = potentialSolution.pop()
        nextState = currentState.newState()
        nextState.setscore()
        delta = (nextState.sc- currentState.sc)*scale
        temperature = (maxtemp - countTemp)
        if (delta > 0):
            currentState = nextState
            countTemp += 1
            countTime = 0
        else:
            
            
            probabilitytaken = math.exp(delta/temperature)

            if (random.random() < probabilitytaken):
                currentState = nextState
                countTemp += 1

            countTime += 1

        potentialSolution.append(currentState)
        if ((countTime == 100)or(temperature <= 1)):
            countTemp = 0
            temperature = maxtemp
            countTime = 0
            countSolution +=1
            currentState = state.shuffle()
            currentState.setscore()
            potentialSolution.append(currentState)

    bestSolution = potentialSolution[0]
    for i in range(0, len(potentialSolution)):
        if (potentialSolution[i].score() > bestSolution.score()):
            bestSolution = potentialSolution[i]

    return bestSolution

File: C-version/method.c
#include "state.h"
#include "helper.h"
#include <sys/time.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

struct State hillclimbing(struct State state, int usecs) {
    struct State current_state;
    new_state(&current_state, &state, 0);
    struct State next_state;

    // variable sized array implementation
    // (because i'm too far gone to use std::vector)
    int max_size = 10000;
    struct State *potential_solution = calloc(max_size, sizeof(struct State));
    int solutions = 0;
    potential_solution[0] = current_state;
    solutions++;

    int count = 0;
    struct timeval beg;
    struct timeval end;
    gettimeofday(&beg, 0);
    do {
        current_state = potential_solution[solutions-1];
        new_state(&next_state, &current_state, 1);
        if (score(&next_state) > score(&current_state)) {
            delete_state(&current_state);
            current_state = next_state;
            count = 0;
        }
        else {
            delete_state(&next_state);
            count++;
        }
        potential_solution[solutions-1] = current_state;
        if (count == 100) {
            count = 0;
            solutions++;
            new_state(&current_state, &state, 0);
            shuffle(&current_state);
            potential_solution[solutions-1] = current_state;
        }
        // expand potential solution
        if (solutions == max_size - 10) {
            max_size += 10000;
            potential_solution = realloc(potential_solution, sizeof(struct State) * max_size);
        }
        gettimeofday(&end, 0);
    } while (elapsed_useconds(beg, end) < usecs);
    for (int i = 0; i < solutions; i++) {
        if (score(&current_state) < score(&potential_solution[i]))
            current_state = potential_solution[i];
    }
    free(potential_solution);
    return current_state;
}

struct State annealing(struct State state, double maxtemp, int usecs) {
    struct State current_state;
    new_state(&current_state, &state, 0);
    struct State next_state;

    // variable sized array implementation
    // (because i'm too far gone to use std::vector)
    int max_size = 10000;
    struct State *potential_solution = calloc(max_size, sizeof(struct State));
    int solutions = 0;
    potential_solution[0] = current_state;
    solutions++;

    int count = 0;
    double counttemp = 0;
    double temperature;
    struct timeval beg;
    struct timeval end;
    gettimeofday(&beg, 0);
    do {
        current_state = potential_solution[solutions-1];
        new_state(&next_state, &current_state, 1);
        double delta = score(&next_state) > score(&current_state);
        if (delta > 0) {
            delete_state(&current_state);
            current_state = next_state;
            count = 0;
        }
        else {
            temperature = (double) maxtemp - counttemp;
            if (temperature < 0) {
                temperature = 0.1;
            }
            double probabilitytaken = exp(delta/temperature);
            if (decision(probabilitytaken)) {
                delete_state(&current_state);
                current_state = next_state;
                counttemp += 1;
            }
            else delete_state(&next_state);
            count++;
        }
        potential_solution[solutions-1] = current_state;
        if (count == 100) {
            count = 0;
            solutions++;
            new_state(&current_state, &state, 0);
            shuffle(&current_state);
            potential_solution[solutions-1] = current_state;
        }
        // expand potential solution
        if (solutions == max_size - 10) {
            max_size += 10000;
            potential_solution = realloc(potential_solution, sizeof(struct State) * max_size);
        }
        gettimeofday(&end, 0);
    } while (elapsed_useconds(beg, end) < usecs);
    for (int i = 0; i < solutions; i++) {
        if (score(&current_state) < score(&potential_solution[i]))
            current_state = potential_solution[i];
    }
    free(potential_solution);
    return current_state;
}
